core/src/main/scala/org/apache/spark/util/JsonProtocol.scala

```diff
@@ -731,183 +925,180 @@ private[spark] object JsonProtocol {
+    val executorInfo = executorInfoFromJson(json.get("Executor Info"))
     SparkListenerExecutorAdded(time, executorId, executorInfo)
   }
 
-  def executorRemovedFromJson(json: JValue): SparkListenerExecutorRemoved = {
-    val time = (json \ "Timestamp").extract[Long]
-    val executorId = (json \ "Executor ID").extract[String]
-    val reason = (json \ "Removed Reason").extract[String]
+  def executorRemovedFromJson(json: JsonNode): SparkListenerExecutorRemoved = {
+    val time = json.get("Timestamp").longValue
+    val executorId = json.get("Executor ID").textValue
+    val reason = json.get("Removed Reason").textValue
     SparkListenerExecutorRemoved(time, executorId, reason)
   }
 
-  def logStartFromJson(json: JValue): SparkListenerLogStart = {
-    val sparkVersion = (json \ "Spark Version").extract[String]
+  def logStartFromJson(json: JsonNode): SparkListenerLogStart = {
+    val sparkVersion = json.get("Spark Version").textValue
     SparkListenerLogStart(sparkVersion)
   }
 
-  def executorMetricsUpdateFromJson(json: JValue): SparkListenerExecutorMetricsUpdate = {
-    val execInfo = (json \ "Executor ID").extract[String]
-    val accumUpdates = (json \ "Metrics Updated").extract[List[JValue]].map { json =>
-      val taskId = (json \ "Task ID").extract[Long]
-      val stageId = (json \ "Stage ID").extract[Int]
-      val stageAttemptId = (json \ "Stage Attempt ID").extract[Int]
+  def executorMetricsUpdateFromJson(json: JsonNode): SparkListenerExecutorMetricsUpdate = {
+    val execInfo = json.get("Executor ID").textValue
+    val accumUpdates = json.get("Metrics Updated").elements.asScala.map { json =>
+      val taskId = json.get("Task ID").longValue
+      val stageId = json.get("Stage ID").intValue
+      val stageAttemptId = json.get("Stage Attempt ID").intValue
       val updates =
-        (json \ "Accumulator Updates").extract[List[JValue]].map(accumulableInfoFromJson)
+        json.get("Accumulator Updates").elements.asScala.map(accumulableInfoFromJson).toArray.toSeq
       (taskId, stageId, stageAttemptId, updates)
-    }
-    val executorUpdates = (json \ "Executor Metrics Updated") match {
-      case JNothing => Map.empty[(Int, Int), ExecutorMetrics]
-      case value: JValue => value.extract[List[JValue]].map { json =>
-        val stageId = (json \ "Stage ID").extract[Int]
-        val stageAttemptId = (json \ "Stage Attempt ID").extract[Int]
-        val executorMetrics = executorMetricsFromJson(json \ "Executor Metrics")
+    }.toArray.toSeq
+    val executorUpdates = jsonOption(json.get("Executor Metrics Updated")).map { value =>
+      value.elements.asScala.map { json =>
+        val stageId = json.get("Stage ID").intValue
+        val stageAttemptId = json.get("Stage Attempt ID").intValue
+        val executorMetrics = executorMetricsFromJson(json.get("Executor Metrics"))
         ((stageId, stageAttemptId) -> executorMetrics)
       }.toMap
-    }
+    }.getOrElse(Map.empty)
```

Comment:
seems should be `Map.empty[(Int, Int), ExecutorMetrics]`, otherwise compilation will fail with `Scala-2.13`:

```
[error] /home/runner/work/spark/spark/core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:1061:64: type mismatch;
[error]  found   : scala.collection.Map[_1,org.apache.spark.executor.ExecutorMetrics] where type _1 <: (Int, Int)
[error]  required: scala.collection.Map[(Int, Int),org.apache.spark.executor.ExecutorMetrics]
[error] Note: _1 <: (Int, Int), but trait Map is invariant in type K.
[error] You may wish to investigate a wildcard type such as `_ <: (Int, Int)`. (SLS 3.2.10)
[error]     SparkListenerExecutorMetricsUpdate(execInfo, accumUpdates, executorUpdates)
[error]                                                                ^
[error] one error found
[error] (core / Compile / compileIncremental) Compilation failed
[error] Total time: 136 s (02:16), completed Jun 16, 2022 2:07:49 AM
```